### PR TITLE
2873 version field in agent plugin manifest

### LINKED ISSUE #2873: Add a `version` field to `AgentPluginManifest`
## Description

Plugins should have their own versions so that they can be updated and versioned independently of the core system. Since plugins are not expected to be installable Python (pip) packages, they should use [semantic](https://semver.org/) versioning, **not** [PEP 440](https://peps.python.org/pep-0440/) versioning.

### Resources
https://python-semver.readthedocs.io/en/latest/advanced/combine-pydantic-and-semver.html

## Tasks

- [ ] Add a version field to the `AgentPluginManifest` (0.25d) @ilija-lazoroski 
- [x] Add version field to all manifests (0d) @ilija-lazoroski 
    - [x] hadoop plugin
    - [x] hard-coded exploiters
    - [x] mock plugins

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #2873 .

Add any further explanations here.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/common/agent_plugins/agent_plugin.py [hunk 1]
```diff
@@ -25,6 +26,11 @@ class AgentPlugin(InfectionMonkeyBaseModel):
...
 
-    class Config:
+    class Config(InfectionMonkeyModelConfig):
         # b64encode() returns bytes, so we call decode() to transform bytes to str
-        json_encoders = {bytes: lambda byte_field: b64encode(byte_field).decode()}
+        # PluginVersion needs to have json encoder also at the top level,
+        # as pydantic is not able to inherit AgentPluginManifest json encoders
+        json_encoders = {
+            bytes: lambda byte_field: b64encode(byte_field).decode(),
+            PluginVersion: lambda v: str(v),
```
Comment (cakekoa) [inline monkey/common/agent_plugins/agent_plugin.py:35]: Is this needed, or is the encoder in `AgentPluginManifest` sufficient?

File: monkey/tests/unit_tests/common/agent_plugins/test_agent_plugin_manifest.py [hunk 2]
```diff
@@ -97,3 +101,12 @@ def test_agent_plugin_manifest__remediation_suggestion():
...
 
     assert agent_manifest_object.remediation_suggestion == remediation_suggestion
+
+
+def test_agent_plugin_manifest__version():
+    import pydantic
+
+    from common.types import PluginVersion
+
+    class MyModel(pydantic.BaseModel):
```
Comment (cakekoa) [inline monkey/tests/unit_tests/common/agent_plugins/test_agent_plugin_manifest.py:111]: This test looks incomplete

**Review by cakekoa — COMMENTED** [review]
Should I be concerned that a few items are missing from the `Pipfile.lock` files? I'm not sure what to look for there.

File: monkey/common/agent_plugins/agent_plugin.py (hunk 1 above)
Comment (ilija-lazoroski) [reply monkey/common/agent_plugins/agent_plugin.py:35]: This is needed, as `AgentPlugin` can't really use the encoder from `AgentPluginManifest`

**Comment by mssalvatore** [comment]
> Should I be concerned that a few items are missing from the Pipfile.lock files? I'm not sure what to look for there

Yeah, that's a problem. If you install a new dependency on Linux, you need to run `pip install --keep-outdated NEW_DEPENDENCY`. Otherwise it removes windows-specific dependencies. Pipenv is... well, whatever.

File: monkey/common/agent_plugins/agent_plugin.py (hunk 1 above)
Comment (VakarisZ) [reply monkey/common/agent_plugins/agent_plugin.py:35]: It's strange that it's so hard to find information about this 🤔 at least do:
```
json_encoders.update(AgentPluginManifest.Config.json_encoders)
```

Comment (VakarisZ) [reply monkey/common/agent_plugins/agent_plugin.py:35]: I think this will be supported in the v2 of pydantic

**Review by VakarisZ — CHANGES_REQUESTED** [review]
Where are the tests that ensure failure if the version is malformed?

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**